core/src/layers/throttle.rs

```diff
@@ -0,0 +1,315 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::io::SeekFrom;
+use std::num::NonZeroU32;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::thread;
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use governor::clock::{Clock, DefaultClock};
+use governor::middleware::NoOpMiddleware;
+use governor::state::{InMemoryState, NotKeyed};
+use governor::{NegativeMultiDecision, Quota, RateLimiter};
+
+use crate::raw::oio::Streamer;
+use crate::raw::*;
+use crate::*;
+
+/// Add a bandwidth rate limiter to the underlying services.
+///
+/// # Throttle
+///
+/// There are several algorithms when it come to rate limiting techniques.
+/// This throttle layer uses Generic Cell Rate Algorithm (GCRA) provided by
+/// [Governor](https://docs.rs/governor/latest/governor/index.html).
+/// By setting the `replenish_byte_rate` and `max_burst_byte`, we can control the byte flow rate of underlying services.
+///
+/// # Note
+///
+/// When setting the ThrottleLayer, always consider the largest possible operation size as the burst size,
+/// as **the burst size should be larger than any possible byte length to allow it to pass through**.
+///
+/// Read more about [Quota](https://docs.rs/governor/latest/governor/struct.Quota.html#examples)
+///
+/// # Examples
+///
+/// This example limits bandwidth to 10 KiB/s and burst size to 10 MiB.
+/// ```
+/// use anyhow::Result;
+/// use opendal::layers::ThrottleLayer;
+/// use opendal::services;
+/// use opendal::Operator;
+/// use opendal::Scheme;
+///
+/// let _ = Operator::new(services::Memory::default())
+///     .expect("must init")
+///     .layer(ThrottleLayer::new(10 * 1024, 10000 * 1024))
+///     .finish();
+/// ```
+#[derive(Clone)]
+pub struct ThrottleLayer {
+    replenish_byte_rate: u32,
+    max_burst_byte: u32,
+}
+
+impl ThrottleLayer {
+    /// Quota takes replenish_byte_rate and max_burst_byte as NonZeroU32 types, so setting them to <= 0 will trigger a panic.
+    /// NonZeroU32 can store an integer from 1 to 4,294,967,295 (2^32 - 1).
+    pub fn new(replenish_byte_rate: u32, max_burst_byte: u32) -> Self {
+        assert!(replenish_byte_rate > 0);
+        assert!(max_burst_byte > 0);
+        Self {
+            replenish_byte_rate,
+            max_burst_byte,
+        }
+    }
+}
+
+impl<A: Accessor> Layer<A> for ThrottleLayer {
+    type LayeredAccessor = ThrottleAccessor<A>;
+
+    fn layer(&self, accessor: A) -> Self::LayeredAccessor {
+        let rate_limiter = Arc::new(RateLimiter::direct(
+            Quota::per_second(NonZeroU32::new(self.replenish_byte_rate).unwrap())
```

Comment:
We can store `NonZeroU32` in `ThrottleLayer` directly so since we already checked them.

Reply:
Done